### ISSUE TITLE
JOINTS_0 pointing to what exactly?

### ISSUE THREAD

**Issue by QTM3** [KhronosGroup/glTF#2141]
I am new to GLTF and have run into an issue that I could not find a solution to. In different GLTF files JOINTS_0 either point to jointIDs in the global set of Joints i.e. model.nodes or it points to skin.joints which then point to model.nodes, and there is no way of knowing which one is the correct one.

Following assumes "joints" points to the "JOINTS_0" buffer and "jointStride" is properly computed from ByteStride and ComponentSizeInBytes. "vid" is the vertex id.

For example, in some cases THIS gives the correct results (which I think is the GLTF standard)
for (int i = 0; i < 4; i++)
{
int jntid = joints[vid * jointStride + i];
vertex.jointid[i] = skin.joints[jntid]; 
}

Whereas with some other files THIS gives the correct results
for (int i = 0; i < 4; i++)
{
vertex.jointid[i] = joints[vid * jointStride + i]; 
}

How can one know which one to use and when? I am using tinygltf for reading the GLTF files. I am sure I am doing something wrong but I could not find a solution, thanks in advance for any suggestions.

**Comment by javagl** [comment]
> Any glTF file with JOINTS_N attributes referencing nodes not in the skin.joints list would be invalid and should fail the [validation suite](https://github.com/KhronosGroup/glTF-Validator/).

An aside: I think that in some cases, the validator may not even be able to detect whether something is wrong here. As QTM3 mentioned, there is no way of telling whether the indices are correct. 

If `JOINTS_0` contains `[0,1,2,3]`, then it is not obvious whether this refers to some `skin.joints = [3,2,1,0]` or to the top-level `nodes` (even though the latter would be wrong, and a bug in the exporter). In the worst case, this would not yield a validation error, but only a messed-up result at runtime. (Unless, coincidentally, `skin.joints = [0,1,2,3]`, in which case this error could remain entirely undetected...)

**Comment by QTM3** [comment]
@donmccurdy thank you for the response. I think you mean to say the 1st code is the correct one as that is the one referencing the skin.joints list. I tested the GLB files on the online validator https://github.khronos.org/glTF-Validator/ and indeed the one that uses the 2nd code (directly referencing the root.nodes) has errors, is it because of this, I cannot say. However blender seems to get right past these issues and correctly displays both files, thus I was wondering if there is a property which is set if direct root.nodes are referenced by JOINTS_N. 

As @javagl pointed out there is very little we can infer if the indices are not right, only a mangled mess is displayed. I tried using Assimp for the same and even Assimp could not load the direct referencing file. Both files are available at the following links if interested.

Working good file : https://www.mediafire.com/file/s55wa120ypbro9n/Working.glb/file
Problematic bad file: https://www.mediafire.com/file/b3e6y690etr5kb4/Probmematic.glb/file

**Comment by donmccurdy** [comment]
Oh I see — you're intending to store an index in into the _root_ nodes list as `vertex.jointid`?  I don't know if you want to upload the positions of all nodes to your GPU (assuming you're writing a viewer?) but I understand what you mean. E.g. in three.js we generate a texture containing the transforms of only those nodes used as joints.

Maybe for clarity I'd rephrase it this way:

```c
for (int i = 0; i < 4; i++)
{
  int jntid = joints[vid * jointStride + i];
  vertex.jointNodeId[i] = skin.joints[jntid];
}
```

**Comment by QTM3** [comment]
True I am reading the file and storing the data in my own structures which are then parsed and uploaded to the GPU to blend the skinning matrices and perform various other operations. For this I use tinyGLTF and Assimp. Both are having trouble with direct referencing JOINTS_N. For now I have allowed the user to select a checkbox if the mesh is mangled to switch between the two approaches, but this does not seem like a good solution as both Blender and the Windows 3D viewer can properly display both files. It seems like I am missing something here.

**Comment by scurest** [comment]
It doesn't make any sense not to use the skin because you need the inverse binds matrices to do skinning.

**Comment by QTM3** [comment]
> It doesn't make any sense not to use the skin because you need the inverse binds matrices to do skinning.

Yes, that is true, it does not make sense to directly address root.nodes, but that is what I am observing in these files. I may be doing something wrong, I may be assuming something about the order of joints in the root.nodes. This line in Section 3.7.3.2 of the GLTF 2.0 ( https://www.khronos.org/registry/glTF/specs/2.0/glTF-2.0.html ) specification might suggests the correct order :

> Implementation Note
> A node object does not specify whether it is a joint. Client implementations may need to traverse the skins array first, marking each joint node.

The skin.joints may be pointing to an array which only contains joints from the root.nodes, rather than directly pointing to the root.nodes. I will try and implement this and get back with the results.
